lock-files.sh

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env bash
+#
+# Create and manipulate the lock files.
+#
+# We use lock files `Cargo-recent.lock` and `Cargo-minimal.lock` to pin
+# dependencies and check specific versions in CI.
+#
+# Shellcheck can't search dynamic paths
+# shellcheck source=/dev/null
+
+set -u
+
+main() {
+    set_globals
+    assert_cmds
+    handle_command_line_args "$@"
+}
+
+set_globals() {
+    # The directory of this script.
+    script_dir=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+
+    # Used below by `say`.
+    script=${0##*/}
+
+    # Used below by `verbose_say`.
+    flag_verbose=false
+
+    # Sourcing this file requires `flag_verbose` to be set.
+    . "$script_dir/stdlib.sh"
+
+    # Environment sanity checks
+    assert_nz "$HOME" "\$HOME is undefined"
+    assert_nz "$0" "\$0 is undefined"
+
+    # Make all cargo invocations verbose.
+    export CARGO_TERM_VERBOSE=true
+
+    recent="Cargo-recent.lock"
+    minimal="Cargo-minimal.lock"
+}
+
+handle_command_line_args() {
+    local _help=false
+    local _create=""
+    local _update=""
+
+    local _arg
+    for _arg in "$@"; do
+        case "${_arg%%=*}" in
+            -c | --create )
+                create
+                ;;
+
+            -u | --update )
+                update
+                ;;
+
+            -h | --help )
+                _help=true
+                ;;
+
+            --verbose)
+                # verbose is a global flag
+                flag_verbose=true
+                ;;
+
+            *)
+                echo "Unknown argument '$_arg', displaying usage:"
+                echo "${_arg%%=*}"
+                _help=true
+                ;;
+
+        esac
+
+    done
+
+    if [ "$_help" = true ]; then
+        print_help
+        exit 0
+    fi
+    verbose_say "Your git index will now be dirty if lock file update is required"
+}
+
+# Creates the minimal and recent lock files.
+#
+# If this function fails you may want to be lazy and just duplicate `Cargo.lock`
+# as the minimal and recent lock files.
+create() {
+    need_nightly
+
+    # Make a best effort at creating a minimal lock file.
+    rm --force Cargo.lock
+    cargo +nightly check --all-features -Z minimal-versions
```

Comment:
In 65a74a14d7f5cddc79e77a44289460e8aec97a82:

It seems like this script needs this line to fail in order to be useful. But I believe this fails on every single one of our repos becaue our minimal dependencies themselves don't specify their minimal dependencies correctly.

Though I guess I haven't tried it for some of the new leaf crates..

Reply:
> It seems like this script needs this line to fail in order to be useful. 

What do you mean by this? My main idea behind this script was that if this succeeds then win and one doesn't have to have the `-Z` command in shell history or in mind. And if it this fails the script documents my lazy 'just duplicate recent'. But since the script lives in an obscure repo is this even useful, I don't know?

I used `-Z minimal-versions` recently and it worked (maybe in `bech32`, but not sure) to my surprise.

Reply:
> I used `-Z minimal-versions` recently and it worked, to my surprise (maybe in `bech32`, but not sure).

Ooo, nice.

I think the MSRV update to 1.63 solved a lot of our dependency compatibility issues.

>And if it this fails the script documents my lazy 'just duplicate recent'

I don't think it's too useful as documentation because it's inside a script in a fairly obscure repo.

I also think we should move away from "duplicate the recent lockfile" to "just don't have a minimal lockfile". This will clearly signal which repos still need somebody to do the hard work of running `-Z minimal` then patching everything up.

Reply:
I guess, if the script were to try using -Zminimal, and on failure, output a warning and not produce the lockfile, that'd be pretty cool.

I guess we also need to check that the "recent" lockfile produced by `cargo update` works with MSRV. Again, in my experience it almost never does. And we need to hard fail if neither lockfile can be produced.

Reply:
Good ideas, will re-spin.

Reply:
Currently the script doesn't fail correctly in workspace because `cargo check` in a workspace does not exit with non-zero if the last crate builds successfully, I couldn't find a way to make it do so.

> I also think we should move away from "duplicate the recent lockfile" to "just don't have a minimal lockfile".

Leaving this for another day.